daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala

```diff
@@ -35,53 +35,70 @@ private[lf] final class CommandPreprocessor(
     speedy.Command.Create(templateId, arg)
   }
 
-  def unsafePreprocessExercise(
-      identifier: Ref.Identifier,
+  def unsafePreprocessLenientExercise(
+      templateId: Ref.Identifier,
       contractId: Value.ContractId,
       choiceId: Ref.ChoiceName,
       argument: Value,
-  ): speedy.Command = {
-    import language.PackageInterface.ChoiceInfo
-
-    val cid = valueTranslator.unsafeTranslateCid(contractId)
-    def command(
-        choice: Ast.TemplateChoiceSignature,
-        toSpeedyCommand: speedy.SValue => speedy.Command,
-    ) = {
-      val arg = valueTranslator.unsafeTranslateValue(choice.argBinder._2, argument)
-      toSpeedyCommand(arg)
+  ): speedy.Command =
+    handleLookup(interface.lookupLenientChoice(templateId, choiceId)) match {
+      case PackageInterface.ChoiceInfo.Template(choice) =>
+        speedy.Command.ExerciseTemplate(
+          templateId = templateId,
+          contractId = valueTranslator.unsafeTranslateCid(contractId),
+          choiceId = choiceId,
+          argument = valueTranslator.unsafeTranslateValue(choice.argBinder._2, argument),
+        )
+      case PackageInterface.ChoiceInfo.Inherited(ifaceId, choice) =>
+        speedy.Command.ExerciseInterface(
+          interfaceId = ifaceId,
+          contractId = valueTranslator.unsafeTranslateCid(contractId),
+          choiceId = choiceId,
+          argument = valueTranslator.unsafeTranslateValue(choice.argBinder._2, argument),
+        )
```

Comment:
This should be an `speedy.Command.ExercisyByInterface` I think. (The difference is that this adds a template id check in the choice guard.)